### PR TITLE
Balk at using Kubernetes versions too new to be supported

### PR DESCRIPTION
**Pull request by johngmyers**

### CODE REVIEW

**Comment by johngmyers** [comment]
e2e test is failing because it is trying to use a k8s 1.19 alpha, which the added code prevents.

File: upup/pkg/fi/cloudup/apply_cluster.go [hunk 1]
```diff
@@ -1053,6 +1053,29 @@ func (c *ApplyClusterCmd) validateKubernetesVersion() error {
...
 
+	kopsVersion, err := semver.Parse(kopsbase.KOPS_RELEASE_VERSION)
+	if err != nil {
+		klog.Warningf("unable to parse kops version %q", kopsVersion)
+	} else {
+		tooNewVersion := kopsVersion
+		tooNewVersion.Minor += 1
+		tooNewVersion.Pre = nil
+		tooNewVersion.Build = nil
+		if util.IsKubernetesGTE(tooNewVersion.String(), *parsed) {
```
Comment (justinsb) [inline upup/pkg/fi/cloudup/apply_cluster.go]: I wonder about GT vs GTE.  I think this means that e.g. kops 1.18 won't be able to run k8s 1.19 alpha.1 ?

It's probably fine but you know me - super careful!

Shall we discuss in office hours?

**Comment by justinsb** [comment]
Another option we could do is to set these guidelines in the channel file (`channels/stable`).  Though I do agree this is a nice way to do it - thanks for starting us off here!

File: upup/pkg/fi/cloudup/apply_cluster.go (hunk 1 above)
Comment (johngmyers) [reply upup/pkg/fi/cloudup/apply_cluster.go]: I intended for kops 1.18 to balk at running k8s 1.19.0-alpha.1 (unless the environment variable was set)

**Comment by johngmyers** [comment]
Would we expect to change this guideline post-release? If not, I don't see the benefit of setting this in the channel file.

**Comment by hakman** [comment]
Any special reason this is desired? I would rather not have this behaviour at all, especially not another env var. Anyone that chooses a newer Kubernetes version does this already at its own risk by setting it manually in the cluster spec.

File: upup/pkg/fi/cloudup/apply_cluster.go (hunk 1 above)
Comment (justinsb) [reply upup/pkg/fi/cloudup/apply_cluster.go]: Yeah, I saw that after I commented :-)

It does mean that whatever tests we're still running against k8s HEAD will break at some random point in the cycle when k/k cuts the next alpha.  OTOH that might be a good promopt for us to cut the branch for the next release - though it is typically pretty early in the cycle meaning we'd have to do a _lot_ of cherry picks.

Comment (rifelpet) [reply upup/pkg/fi/cloudup/apply_cluster.go]: FWIW this test-infra PR will fix the prow tests: https://github.com/kubernetes/test-infra/pull/16678/files

**Comment by johngmyers** [comment]
I've seen several issues filed by users who were using older versions of kops than the Kubernetes they were running, with no idea that wasn't supported. It would be to the users' benefit if kops declined to perform actions that are so risky to merit an explicit disclaimer of support.

**Comment by hakman** [comment]
We have to find a balance between users not reading any docs and develoment pain. Having env vars required to test newer k8s versions is not pleasant. I think a good balance would be to keep only the warning.

**Comment by johngmyers** [comment]
Setting an environment variable is easy, particularly compared to having to host your own assets. I believe this is a case where the needs of the users outweigh convenience to developers. The environment variable override is the pattern established by the code handling k8s versions marked as obsolete in the channel.

**Comment by hakman** [comment]
> The environment variable override is the pattern established by the code handling k8s versions marked as obsolete in the channel.

This is an old piece of code that will probably be replaced by the new approach of not supporting old releases in 1.19 or 1.20.

I think this is something on which we will disagree. From my point of view, env vars are unfriendly. Even the message is annoying to see every time you run a command.

This is not a tool for the casual user. If someone doesn't spend the time to even check the compatibility matrix, there is a very slim chance that it will be able to host its own cluster.
There will be similar questions when the user will not read the release notes and so on.

**Comment by olemarkus** [comment]
FWIW the place I have seen problems here is where one admin forgets to update kops locally and accidentally causes problems. Say what you want about such mistakes, but it does happen.

**Comment by hakman** [comment]
> FWIW the place I have seen problems here is where one admin forgets to update kops locally and accidentally causes problems. Say what you want about such mistakes, but it does happen.

OK. Can we do it in such a way that it only applies to official releases?

**Comment by johngmyers** [comment]
Heh, our wrapper around kops puts its version in an annotation on the cluster spec. It then requires a switch to run an older version on that cluster.

**Comment by rifelpet** [comment]
we decided during last office hours that this is good to merge. developers can set the env var in their shell, and CI jobs will do the same according to the test-infra PR linked above.

/lgtm
/approve

**Pull request merged**